src/passes/PreJSRewriter.cpp

```diff
@@ -0,0 +1,150 @@
+/*
+ * Copyright 2021 WebAssembly Community Group participants
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <pass.h>
+#include <wasm.h>
+
+#include "wasm-builder.h"
+#include <ir/literal-utils.h>
+#include <ir/localize.h>
+
+namespace wasm {
+
+struct PreJSRewriterPass : public WalkerPass<PostWalker<PreJSRewriterPass>> {
+  std::unique_ptr<Builder> builder;
+
+  bool isFunctionParallel() override { return false; }
+
+  Pass* create() override { return new PreJSRewriterPass; }
+
+  void doWalkFunction(Function* func) {
+    if (!builder) {
+      builder = make_unique<Builder>(*getModule());
+    }
+    super::doWalkFunction(func);
+  }
+
+  void visitBinary(Binary* curr) {
+    switch (curr->op) {
+      case EqInt32:
+      case EqInt64:
+        // Rewrite popcnt(x) == 1   ==>   !(!x | (x & (x - 1))
+        if (auto* lhs = curr->left->dynCast<Unary>()) {
+          if (auto* rhs = curr->right->dynCast<Const>()) {
+            if ((lhs->op == PopcntInt32 || lhs->op == PopcntInt64) &&
+                rhs->value.getInteger() == 1LL) {
+              rewritePopcntEqualOne(curr);
+            }
+          }
+        }
+        break;
+
+      case CopySignFloat32:
+      case CopySignFloat64:
+        rewriteCopysign(curr);
+        break;
+
+      default:
+        break;
+    }
+  }
+
+  void rewriteCopysign(Binary* curr) {
```

Comment:
It would be good to have a comment showing what transformation is being done here, too.